stratum/hal/bin/np4intel/update_config.py

```diff
@@ -0,0 +1,157 @@
+            google.protobuf.text_format.Merge(p4info_f.read(), config.p4info)
+        device_config = build_bmv2_config(args.json)
+        config.p4_device_config = device_config.SerializeToString()
+        request.action = p4runtime_pb2.SetForwardingPipelineConfigRequest.VERIFY_AND_COMMIT
+        try:
+            self.stub.SetForwardingPipelineConfig(request)
+        except Exception as e:
+            print "Error during SetForwardingPipelineConfig"
+            print str(e)
+            return False
+        return True
+
+    def set_up_stream(self):
+        self.stream_out_q = Queue.Queue()
+        self.stream_in_q = Queue.Queue()
+        def stream_req_iterator():
+            while True:
+                p = self.stream_out_q.get()
+                if p is None:
+                    break
+                yield p
+
+        def stream_recv(stream):
+            for p in stream:
+                self.stream_in_q.put(p)
+
+        self.stream = self.stub.StreamChannel(stream_req_iterator())
+        self.stream_recv_thread = threading.Thread(
+            target=stream_recv, args=(self.stream,))
+        self.stream_recv_thread.start()
+
+        self.handshake()
+
+    def handshake(self):
+        req = p4runtime_pb2.StreamMessageRequest()
+        arbitration = req.arbitration
+        arbitration.device_id = self.device_id
+        election_id = arbitration.election_id
+        election_id.high = 0
+        election_id.low = 1
+        self.stream_out_q.put(req)
+
+        rep = self.get_stream_packet("arbitration", timeout=3)
+        if rep is None:
+            print "Failed to establish handshake"
+
+    def tear_down_stream(self):
+        self.stream_out_q.put(None)
+        self.stream_recv_thread.join()
+
+    def get_packet_in(self, timeout=3):
+        msg = self.get_stream_packet("packet", timeout)
+        if msg is None:
+            print "Packet in not received"
+        else:
+            return msg.packet
+
+    def get_stream_packet(self, type_, timeout=1):
+        start = time.time()
+        try:
+            while True:
+                remaining = timeout - (time.time() - start)
+                if remaining < 0:
+                    break
+                msg = self.stream_in_q.get(timeout=remaining)
+                if not msg.HasField(type_):
+                    continue
+                return msg
+        except:  # timeout expired
+            pass
+        return None
+
+    def send_packet_out(self, packet):
+        packet_out_req = p4runtime_pb2.StreamMessageRequest()
+        packet_out_req.packet.CopyFrom(packet)
+        self.stream_out_q.put(packet_out_req)
+
+if __name__ == '__main__':
+    test = Test()
+    test.set_up_stream()
+    test.update_config()
+    if args.loopback:
+        packet_out = p4runtime_pb2.PacketOut()
+        packet_out.payload = "\xab" * 100
+        test.send_packet_out(packet_out)
+        test.get_packet_in()
+    test.tear_down_stream()
```

Comment:
Is this file specific to NP4? If not, tools like this are better stored out side of Stratum.

Reply:
you're right, I've removed it, along with a couple of others that weren't needed

Reply:
removed